File: projector/deno/projector_test.ts
import { assertEquals } from "jsr:@std/assert";
import { Operation } from "./config.ts";
import { type Data, Projector } from "./projector.ts";

function createData(): Data {
  return {
    projector: {
      "/": {
        foo: "bar1",
        fem: "is",
      },
      "/foo": {
        "foo": "bar2",
      },
      "/foo/bar": {
        "foo": "bar3",
      },
    },
  };
}

function getProjector(pwd: string, data = createData()): Projector {
  return new Projector({
    args: [],
    operation: Operation.Print,
    pwd,
    configPath: "",
  }, data);
}

Deno.test("getValueAll", () => {
  const projector = getProjector("/foo/bar");

  assertEquals(projector.getValueAll(), {
    "fem": "is",
    "foo": "bar3",
  });
});

Deno.test("getValue", () => {
  let projector = getProjector("/foo/bar");

  assertEquals(projector.getValue("foo"), "bar3");

  projector = getProjector("/foo");

  assertEquals(projector.getValue("foo"), "bar2");

  assertEquals(projector.getValue("fem"), "is");
});

Deno.test("setValue", () => {
  const data = createData();
  let projector = getProjector("/foo/bar", data);

  projector.setValue("foo", "bar4");
  assertEquals(projector.getValue("foo"), "bar4");

  projector.setValue("baz", "foo");
  assertEquals(projector.getValue("baz"), "foo");

  projector.setValue("fem", "is not");
  assertEquals(projector.getValue("fem"), "is not");

  projector = getProjector("/", data);
  assertEquals(projector.getValue("fem"), "is");
});

Deno.test("removeValue", () => {
  const projector = getProjector("/foo/bar");

  assertEquals(projector.getValue("foo"), "bar3");
  projector.removeValue("foo");
  assertEquals(projector.getValue("foo"), "bar2");
});

File: projector/go/pkg/projector/projector_test.go
package projector_test

import (
	"testing"

	"github.com/OsmarJoseph/projector/go/pkg/projector"
)

func getData() *projector.Data {
	return &projector.Data{
		Projector: map[string]map[string]string{
			"/": {
				"foo": "bar1",
				"fem": "is",
			},
			"/foo": {
				"foo": "bar2",
			},
			"/foo/bar": {
				"foo": "bar3",
			},
		},
	}
}

func getProjector(pwd string, data *projector.Data) *projector.Projector {
	return projector.CreateProjector(&projector.Config{
		Args:      []string{},
		Operation: projector.Print,
		Pwd:       pwd,
		Config:    "",
	}, data)
}

func test(t *testing.T, projector *projector.Projector, key, expectedValue string) {
	value, ok := projector.GetValue(key)

	if !ok {
		t.Errorf("Expected to find value foo")
	}

	if value != expectedValue {
		t.Errorf("Expected value to be %v, got %v", expectedValue, value)
	}
}

func TestGetValue(t *testing.T) {
	data := getData()
	projector := getProjector("/foo/bar", data)

	test(t, projector, "foo", "bar3")
	test(t, projector, "fem", "is")
}

func TestSetValue(t *testing.T) {
	data := getData()
	projector := getProjector("/foo/bar", data)

	test(t, projector, "foo", "bar3")
	projector.SetValue("foo", "bar4")
	test(t, projector, "foo", "bar4")

	projector.SetValue("fem", "is2")
	test(t, projector, "fem", "is2")

	projector = getProjector("/", data)
	test(t, projector, "fem", "is")
}

func TestRemoveValue(t *testing.T) {
	data := getData()
	projector := getProjector("/foo/bar", data)

	test(t, projector, "foo", "bar3")
  projector.RemoveValue("foo")
	test(t, projector, "foo", "bar2")

  projector.RemoveValue("fem")

	test(t, projector, "fem", "is")
}
